### PR TITLE
project(ui): confine users to projects and standardize slug usage

### LINKED ISSUE #71: feat(ui): Confine users to projects
### Expected feature

- Add an option on the create/edit user form to restrict ui and projects access for users
- Restrict access to the interface given
@Talanor

### Alternative solutions

_No response_

### Anything else?

_No response_

### Acknowledgements

- [X] This issue is not a duplicate of an existing bug report.
- [X] I have chosen an appropriate title.
- [X] All requested information has been provided properly.

### LINKED ISSUE #173: bug(login): bad redirect after first time login
### Current Behavior

When you've installed reNgine-ng if you refresh an existing page it will fill the `next` parameter with the path.
When you login for the first time, no default project has been created yet, you're redirected to the Default project creation page.
But if the `next` value is not `/` redirect will fail with a 500 error.

### Expected Behavior

If it was the first launch of reNgine-ng `next` should be empty
Also if the URI does not exists or is not authorized we must redirect to `/` not to a ugly 500 error

### Steps To Reproduce

1. Install reNgine-ng
2. Login with the created user and change the `next` parameter to other things than `/`

### Environment

```markdown
- reNgine: 2.1.0
- OS: Debian 12
- Python: 3.10
- Docker Engine: 
- Docker Compose: 
- Browser:
```

### Anything else?

_No response_

### Acknowledgements

- [X] This issue is not a duplicate of an existing bug report.
- [X] I have chosen an appropriate title.
- [X] All requested information has been provided properly.

### PR DESCRIPTION
**Pull request by psyray**
fixes #71 
fixes #173 

I've added the possibility to confine users to projects

## User creation/update with projects to assign
image
image

## Project creation/update with users to assign
image

## If user try to access a non assign project it returns a 403 page
image

## Project display is restricted to their assigned projects for user
image
image

I've also remove the * import for the views.py in favor of defined import
It's a bad practice to use *

- [x] User creation with project assign
- [x] Project creation with user assign
- [x] Permissions test
- [x] Fully test functionalities after * import remove

If someone could also test on his side it could be good

## Summary by Sourcery

Confine users to specific projects and standardize slug usage across the application. Introduce project-based user management, restrict project access, and enhance user interface for project and user management. Refactor code for improved readability and maintainability, and update tests to cover new features.

New Features:
- Introduce user confinement to projects, allowing users to be assigned to specific projects and restricting their access to only those projects.
- Implement project-based user management, enabling the assignment of users to projects during project creation or update.
- Add functionality to return a 403 error page when a user attempts to access a project they are not assigned to.
- Restrict project display to only those projects assigned to the user.
- Introduce a new Project model with a description field and a many-to-many relationship with users.
- Add middleware to manage project access and ensure users can only access projects they are assigned to.

Bug Fixes:
- Fix various URL and path issues by removing unnecessary slug parameters and updating URL patterns.
- Correct error handling in API views to provide more informative error messages.

Enhancements:
- Standardize slug usage across the application to improve URL consistency and readability.
- Refactor views to remove redundant code and improve readability, including the use of helper functions for repeated logic.
- Enhance user interface for project and user management, including the addition of project descriptions and user role management.
- Improve logging for error handling in various views to aid in debugging and monitoring.

Tests:
- Add tests for new user and project management features, including user creation, update, and deletion.
- Update existing tests to accommodate changes in URL patterns and project management logic.

Chores:
- Remove wildcard imports in favor of explicit imports to improve code clarity and maintainability.
- Update templates to use new URL patterns and context variables for improved consistency.

### CODE REVIEW

**Comment by psyray** [comment]
@AnonymousWP Could you test the functionality on your side.
I will test the import * modification

**Comment by 0b3ud** [comment]
I've tested the feature 
Seems to work well

- When I try to access other projects that my user is not assigned to I get a permission denied 

Thus being said on first login for a new user 
The user is automatically redirected to default/dashboard if you have a default project 
Then they're met with a permission denied 

Maybe Users should be redirected to a project they are allowed to access 

But it is a matter of UX 
Other than that all works fine

**Comment by psyray** [comment]
Thanks for testing

> Thus being said on first login for a new user The user is automatically redirected to default/dashboard if you have a default project Then they're met with a permission denied
> Maybe Users should be redirected to a project they are allowed to access So they should click on projects and then go in the project they are assigned to

You're right I will modify it.
I will redirect the user to the first project he is allowed

**Comment by psyray** [comment]
@sourcery-ai review

**Comment by 0b3ud** [comment]
I have retested the PR 
Ux is now better 

- When a user connects to the web interface they are redirected to one of their projects
- Users cannot access other projects that they aren't assigned to 

Thanks awesome work @psyray

**Review by 0b3ud — CHANGES_REQUESTED** [review]
I found a couple of French comments in `web/dashboard/views.py`
Would be awesome if you can translate them to English 
Other than that your it is all fine
Thanks ;)

File: web/dashboard/views.py [hunk 1]
```diff
@@ -254,6 +265,13 @@ def admin_interface_update(request, slug):
                     password=response.get('password')
                 )
                 assign_role(user, response.get('role'))
+                
+                # Ajout des projets
```
[the remark this answers is not in the document]
Comment (psyray) [reply web/dashboard/views.py:269]: Busted, thanks ;)

**Comment by psyray** [comment]
@sourcery-ai review

**Comment by psyray** [comment]
@sourcery-ai review

File: web/dashboard/views.py [hunk 2]
```diff
@@ -3,38 +3,39 @@
...
+    EndPoint, ScanHistory, Subdomain, Vulnerability, ScanActivity,
+    IpAddress, Port, Technology, CveId, CweId, VulnerabilityTags, CountryISO
+)
+from dashboard.models import Project, OpenAiAPIKey, NetlasAPIKey
+from dashboard.forms import ProjectForm
+from reNgine.definitions import PERM_MODIFY_SYSTEM_CONFIGURATIONS, FOUR_OH_FOUR_URL
 
 logger = logging.getLogger(__name__)
 
 def index(request, slug):
```
[the remark this answers is not in the document]
Comment (psyray) [reply web/dashboard/views.py:34]: I will resolve this later when I will refactor all the project

**Comment by psyray** [comment]
You could test and merge
CodeQL send false positive.
I will resolve this when I will do refactoring

**Comment by 0b3ud** [comment]
> Does it need to be re-tested? Considering @0b3ud already tested it.

I will retest it to be sure and then re review the code :)

**Comment by AnonymousWP** [comment]
Then you can approve. @psyray needs to fix merge conflicts, then it can be *squashed and merged* due to many commits (keeps history clean).

**Comment by psyray** [comment]
> Then you can approve. @psyray needs to fix merge conflicts, then it can be **squashed and merged** due to many commits (keeps history clean).

Fixed

**Review by 0b3ud — CHANGES_REQUESTED** [review]
While reviewing the code found a new bug : (
It on admin interface page
image

 When you try to disable a user you get redirected to this page instead of being redirected back to admin interface page
 
image
 
 The feature still works fine, the users are disable and enabled but u need to manually go back to admin interface

**Comment by psyray** [comment]
> When you try to disable a user you get redirected to this page instead of being redirected back to admin interface page
> 
> image
> 
> The feature still works fine, the users are disable and enabled but u need to manually go back to admin interface

Fixed

**Review by 0b3ud — CHANGES_REQUESTED** [review]
Hello :) 
Found another bug on this PR

When you select a projet (landing on the projects dashboard), and u go on another page the project automatically changes to the first project that the user is assigned to. 

Example :
-Choose any project in your list except for the first one
-Click on targets per example 
-Look at the difference in URLS

 Let me know if you need any further details 
 Thanks in advance

**Review by 0b3ud — APPROVED** [review]
The feature works as expected
I am approving the pr

**Comment by AnonymousWP** [comment]
Next time, please squash and merge if it has so many commits.

**Comment by psyray** [comment]
> Next time, please squash and merge if it has so many commits.

Could you explain how to do this, or a link to a doc

**Comment by AnonymousWP** [comment]
> > Next time, please squash and merge if it has so many commits.
> 
> Could you explain how to do this, or a link to a doc

https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/incorporating-changes-from-a-pull-request/about-pull-request-merges#squash-and-merge-your-commits

**Pull request merged**